rclcpp/test/test_mapped_ring_buffer.cpp

```diff
@@ -0,0 +1,154 @@
+  EXPECT_EQ(nullptr, actual);
+}
+
+/*
+Tests push_and_replace with a temporary object.
+*/
+TEST(test_mapped_ring_buffer, temporary_l_value) {
+  rclcpp::mapped_ring_buffer::MappedRingBuffer<char> mrb(2);
+  // Pass in value with temporary object
+  mrb.push_and_replace(1, std::unique_ptr<char>(new char('a')));
+
+  std::unique_ptr<char> actual;
+  mrb.get_copy_at_key(1, actual);
+  EXPECT_EQ('a', *actual);
+
+  mrb.pop_at_key(1, actual);
+  EXPECT_EQ('a', *actual);
+
+  mrb.get_copy_at_key(1, actual);
+  EXPECT_EQ(nullptr, actual);
+}
+
+/*
+Tests normal usage of the mrb.
+*/
+TEST(test_mapped_ring_buffer, nominal) {
+  rclcpp::mapped_ring_buffer::MappedRingBuffer<char> mrb(2);
+  std::unique_ptr<char> expected(new char('a'));
+  // Store expected value's address for later comparison.
+  char * expected_orig = expected.get();
+
+  EXPECT_FALSE(mrb.push_and_replace(1, expected));
+
+  std::unique_ptr<char> actual;
+  mrb.get_copy_at_key(1, actual);
+  EXPECT_NE(nullptr, actual);
+  if (actual) EXPECT_EQ('a', *actual);
+  EXPECT_NE(expected_orig, actual.get());
+
+  mrb.pop_at_key(1, actual);
+  EXPECT_NE(nullptr, actual);
+  if (actual) EXPECT_EQ('a', *actual);
+  EXPECT_EQ(expected_orig, actual.get());
+
+  mrb.get_copy_at_key(1, actual);
+  EXPECT_EQ(nullptr, actual);
+
+  expected.reset(new char('a'));
+  EXPECT_FALSE(mrb.push_and_replace(1, expected));
+
+  expected.reset(new char('b'));
+  EXPECT_FALSE(mrb.push_and_replace(2, expected));
+
+  expected.reset(new char('c'));
+  EXPECT_TRUE(mrb.push_and_replace(3, expected));
+
+  mrb.get_copy_at_key(1, actual);
+  EXPECT_EQ(nullptr, actual);
+
+  mrb.get_copy_at_key(2, actual);
+  EXPECT_NE(nullptr, actual);
+  if (actual) EXPECT_EQ('b', *actual);
+
+  mrb.get_copy_at_key(3, actual);
+  EXPECT_NE(nullptr, actual);
+  if (actual) EXPECT_EQ('c', *actual);
+}
+
+/*
+Tests get_ownership on a normal mrb.
+*/
+TEST(test_mapped_ring_buffer, get_ownership) {
+  rclcpp::mapped_ring_buffer::MappedRingBuffer<char> mrb(2);
+  std::unique_ptr<char> expected(new char('a'));
+  // Store expected value's address for later comparison.
+  char * expected_orig = expected.get();
+
+  EXPECT_FALSE(mrb.push_and_replace(1, expected));
+
+  std::unique_ptr<char> actual;
+  mrb.get_copy_at_key(1, actual);
+  EXPECT_NE(nullptr, actual);
+  if (actual) EXPECT_EQ('a', *actual);
+  EXPECT_NE(expected_orig, actual.get());
+
+  mrb.get_ownership_at_key(1, actual);
+  EXPECT_NE(nullptr, actual);
+  if (actual) EXPECT_EQ('a', *actual);
+  EXPECT_EQ(expected_orig, actual.get());
+
+  mrb.pop_at_key(1, actual);
+  EXPECT_NE(nullptr, actual);
+  if (actual) EXPECT_EQ('a', *actual);  // The value should be the same.
+  EXPECT_NE(expected_orig, actual.get());  // Even though we pop'ed, we didn't get the original.
+
+  mrb.get_copy_at_key(1, actual);
+  EXPECT_EQ(nullptr, actual);
+}
+
+/*
+Tests the affect of reusing keys (non-unique keys) in a mrb.
+*/
+TEST(test_mapped_ring_buffer, non_unique_keys) {
```

Comment:
Yeah, not sure what do about that. I suppose I could assert that keys are unique on insertion, but that will be a performance hit.